lib/src/models/binned.dart

```diff
@@ -550,56 +608,167 @@ abstract class BinnedChartState<T extends BinnedChart> extends State<T>
   /// and updates the selection controller if available.
   void _onTapUp(TapUpDetails details, AxisPainter axisPainter) {
     focusNode.requestFocus();
+    // Always remove the tooltip first
+    _clearHover();
     // Get the selected bin based on the tap location
     SelectedBin? selectedBin = _getBinOnTap(details.localPosition, axisPainter);
     _updatedBinSelection(selectedBin);
   }
 
-  /// Update the selected bins based on the currently selected bin.
-  void _updatedBinSelection(SelectedBin? selectedBin) {
-    selectedDataPoints = {};
-    if (selectedBin != null) {
-      if (selectedBins == null || selectedBins!.seriesIndex != selectedBin.seriesIndex || !isShifting) {
-        selectedBins = SelectedBinRange(selectedBin.seriesIndex, selectedBin.binIndex, null);
-      } else {
-        if (selectedBins!.startBinIndex == selectedBin.binIndex) {
-          selectedBins = null;
-        } else if (selectedBins!.startBinIndex < selectedBin.binIndex) {
-          selectedBins!.endBinIndex = selectedBin.binIndex;
+  void _selectRangeWithClick(SelectedBin selectedBin) {
+    final seriesIndex = selectedBin.seriesIndex;
+    if (selectedBins!.selectedBins.containsKey(seriesIndex)) {
+      int rangePivot = firstSelectedBin!.binIndex;
+      int endIndex = selectedBin.binIndex;
+
+      if (rangePivot == endIndex) {
+        return;
+      }
+
+      if (lastRangeEnd != null) {
+        if (rangePivot < lastRangeEnd!.binIndex) {
+          for (int i = rangePivot; i <= lastRangeEnd!.binIndex; i++) {
+            selectedBins!.removeBin(seriesIndex, i);
+          }
         } else {
-          selectedBins!.endBinIndex = selectedBins!.startBinIndex;
-          selectedBins!.startBinIndex = selectedBin.binIndex;
+          for (int i = lastRangeEnd!.binIndex; i <= rangePivot; i++) {
+            selectedBins!.removeBin(seriesIndex, i);
+          }
         }
       }
-      if (selectedBins != null) {
-        selectedDataPoints = selectedBins!.getSelectedDataIds(binContainers);
+      lastRangeEnd = selectedBin;
+
+      if (rangePivot < endIndex) {
+        _selectRange(seriesIndex, rangePivot, endIndex, forward: true);
+      } else {
+        _selectRange(seriesIndex, rangePivot, endIndex, forward: false);
       }
     } else {
-      selectedBins = null;
+      selectedBins!.addBin(seriesIndex, selectedBin.binIndex);
     }
+  }
 
-    // Update the selection controller if available
-    if (widget.selectionController != null) {
-      widget.selectionController!.updateSelection(widget.info.id, selectedDataPoints);
+  /// Helper method to select a range of bins
+  void _selectRange(Object seriesIndex, int start, int end, {required bool forward}) {
```

Comment:
Just being nitpicky here, but is `forward` really needed? Instead you could just check if start < end, right?

Reply:
The order in which the set is added to is important in keeping state for any selected ranges, as it captures the direction in which they were selected so they can be unselected properly.